src/udp_communication.cpp

```diff
@@ -0,0 +1,327 @@
+        to_robot_.controls[dof].pressure_agonist = min_pressure_ago;
+        to_robot_.controls[dof].pressure_antagonist = min_pressure_antago;
+        update_pressure(dof, Sign::AGONIST, min_pressure_ago);
+        update_pressure(dof, Sign::ANTAGONIST, min_pressure_antago);
+        send();
+        // keeping track of what are the min and max int pressures
+        // observed in the configuration file.
+        min_pressure_ = std::min(min_pressure_, imin_ago);
+        min_pressure_ = std::min(min_pressure_, imin_antago);
+        max_pressure_ = std::max(max_pressure_, imax_ago);
+        max_pressure_ = std::max(max_pressure_, imax_antago);
+    }
+}
+
+void UDPCommunication::send()
+{
+    // sending a UDP telegram to the robot
+    to_robot_.counter += 1;
+    to_robot_.timestamp_client = time_now().count();
+    ssize_t r = c_send(&socket_, &to_robot_, sizeof(to_robot_));
+    if (r < 0)
+    {
+        std::ostringstream s;
+        s << "failed to send udp package " << strerror(errno) << std::endl;
+        std::string error_message(s.str());
+        throw std::runtime_error(error_message);
+    }
+}
+
+void UDPCommunication::update_pressure(unsigned short dof,
+                                       Sign sign,
+                                       float desired_pressure)
+{
+    if (dof == 3)
+    {
+        // because muscles are inverted for dof 3
+        sign = switch_sign(sign);
+    }
+    // just update the to_robot_ structure, does not send
+    // any telegram
+    if (sign == Sign::AGONIST)
+    {
+        to_robot_.controls[dof].pressure_agonist = desired_pressure;
+    }
+    else
+    {
+        to_robot_.controls[dof].pressure_antagonist = desired_pressure;
+    }
+}
+
+RobotState<NB_DOFS> UDPCommunication::receive()
+{
+    ssize_t r = -1;
+    // We want the latest received UDP package, so we "receive" across
+    // the full queue until we get to an EAGAIN error, which means that
+    // to get a new value, the socket would need to wait (and the socket was set
+    // as non blocking, see the method c_init_socket in this file). When hitting
+    // this error, we know we reached the end of the queue, i.e. the newest data
+    while (r == -1)
+    {
+        r = c_receive(&socket_, &from_robot_, sizeof(from_robot_));
+        if (r < 0 && errno != EAGAIN)
+        {
+            std::ostringstream s;
+            s << "failed to receive udp package: " << errno << " ("
+              << strerror(errno) << ")" << std::endl;
+            std::string error_message(s.str());
+            throw std::runtime_error(error_message);
+        }
+    }
+    // initializing the robot state to return
+    RobotState<NB_DOFS> state{
+        static_cast<int>(from_robot_.data.time_micro_secs_actuator_data),
+        static_cast<int>(from_robot_.data.time_micro_secs_sensor_data),
+        TimePoint(from_robot_.time_micro_secs_udp_send)};
+
+    static const double to_radian = -0.0174533;
```

Comment:
Writing `M_PI/180` instead of the resulting number would be a bit clearer.

Reply:
Also, using `constexpr` instead of `const` may result in better code optimisation (not sure if it really makes a difference, though).